app/controllers/drafts_controller.rb

```diff
@@ -0,0 +1,49 @@
+class DraftsController < ApplicationController
+  requires_login
+
+  skip_before_action :check_xhr, :preload_json
+
+  def index
+    params.require(:username)
+    params.permit(:offset)
+    per_chunk = 30
+
+    user = fetch_user_from_params
+
+    opts = { user_id: user.id,
+             user: user,
+             offset: params[:offset].to_i,
+             limit: per_chunk }
+
+    guardian.ensure_can_see_drafts!(user)
+
+    stream = Draft.stream(opts)
+
+    stream.each do |d|
+      parsedData = JSON.parse(d.data)
+      d.raw = parsedData['reply']
+      if parsedData['categoryId'].present? && !d.category_id.present?
+        d.category_id = parsedData['categoryId']
+      end
+    end
+
+    stream = stream.to_a
+
+    if stream.length == 0
```

Comment:
I don't feel like this logic should be present on the server side. The controller should only be responsibly in generating a standard set of payload and let the client side decides how it should display the drafts based on the different conditions.